editor/editor/mode/visual.js

```diff
@@ -1,13 +1,22 @@
 /**
+ * External dependencies
+ */
+import { findIndex } from 'lodash';
+
+/**
  * Internal dependencies
  */
 import InserterButton from '../../inserter/button';
 
 function Blocks( { blocks, onChange } ) {
-	const onChangeBlock = ( index ) => ( changes ) => {
+	const onChangeBlock = ( uid ) => ( changes ) => {
+		const index = findIndex( blocks, { uid } );
 		const newBlock = {
 			...blocks[ index ],
-			changes
+			attributes: {
+				...blocks[ index ].attributes,
+				...changes
+			}
```

Comment:
Thinking about an idea state shape, it might be nice that when we receive the initial parsed set to transform it into a shape that more easily supports lookup and reordering, like:

```js
{
	blocksByUid: {
		'jz1ajlkq': {
			blockType: 'core/text',
			attributes: {}
		},
		'kzjqoosz': {
			blockType: 'core/image',
			attributes: {}
		}
	},
	blockOrder: [
		'jz1ajlkq',
		'kzjqoosz'
	]
}
```